File: services/ai/routers/intelligence.py
import logging
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Dict, Any, Optional

from models.schemas import TeacherAssistantRequest, TeacherAssistantResponse
from agents.teacher_assistant import generate_teacher_assistant_response

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/intelligence",
    tags=["Educational Intelligence Engine"],
)

# -----------------------------------------------------------------------------
# Pydantic Models
# -----------------------------------------------------------------------------

class StudentPerformanceInput(BaseModel):
    student_id: str
    class_grade: int
    subject: str
    topic: str
    quiz_scores: List[float]

class GrowthEngineMetrics(BaseModel):
    learning_xp: int
    skill_xp: int
    career_xp: int
    innovation_xp: int
    community_xp: int
    growth_level: str

class CareerProfileInput(BaseModel):
    student_id: str
    learning_score: int
    skill_score: int
    interests: List[str]

class ProjectBuilderInput(BaseModel):
    problem_statement: str

class VoiceReportInput(BaseModel):
    student_name: str
    lessons_completed: int
    skills_learned: int
    attendance_percent: int

# -----------------------------------------------------------------------------
# Endpoints
# -----------------------------------------------------------------------------

@router.post("/learning-recovery")
async def generate_learning_recovery(data: StudentPerformanceInput):
    """
    Adaptive Learning Engine
    Analyzes quiz scores to detect weak/strong concepts and generates a recovery plan.
    """
    logger.info(f"Generating learning recovery plan for {data.student_id}")
    
    # Mock AI logic for Phase 3
    weak_concepts = ["Fractions", "Decimals"] if data.subject == "Mathematics" else ["Unknown"]
    
    return {
        "status": "success",
        "analysis": {
            "weak_concepts": weak_concepts,
            "strong_concepts": ["Addition", "Multiplication"],
            "learning_velocity": 0.8
        },
        "recovery_plan": {
            "daily_missions": [f"Practice 10 {c} questions" for c in weak_concepts],
            "recommended_lessons": ["Basic Fractions Video", "Decimal Division Interactive"]
        }
    }


@router.post("/daily-briefing")
async def generate_daily_briefing(data: GrowthEngineMetrics):
    """
    AI Learning Brain
    Acts as Personal Growth Coach, generating a daily Marathi briefing based on 5-pillar stats.
    """
    logger.info("Generating daily briefing")
    
    # Simple logic: Find the lowest XP pillar
    pillars = {
        "अभ्यास (Learning)": data.learning_xp,
        "कौशल्ये (Skills)": data.skill_xp,
        "करिअर (Career)": data.career_xp,
        "नावीन्य (Innovation)": data.innovation_xp,
        "समाज (Community)": data.community_xp
    }
    
    lowest_pillar = min(pillars, key=pillars.get)
    
    briefing_marathi = f"नमस्कार! तू सध्या {data.growth_level} लेव्हल वर आहेस. आज आपण '{lowest_pillar}' मध्ये प्रगती करूया. आजची मिशन पूर्ण करून तुझी लेव्हल वाढव!"
    
    return {
        "status": "success",
        "briefing_text": briefing_marathi,
        "focus_area": lowest_pillar
    }


@router.post("/career-match")
async def generate_career_match(data: CareerProfileInput):
    """
    Career Match Engine
    Matches student profile to optimal careers.
    """
    logger.info(f"Generating career match for {data.student_id}")
    
    return {
        "status": "success",
        "career_matches": [
            {
                "career": "Agriculture Scientist",
                "match_score": 92,
                "reason": "Strong interest in nature and high science learning scores."
            },
            {
                "career": "Drone Pilot",
                "match_score": 85,
                "reason": "Good spatial awareness skills."
            }
        ]
    }


@router.post("/project-builder")
async def generate_project_plan(data: ProjectBuilderInput):
    """
    AI Project Builder
    Generates a structured solution and project plan for a given problem.
    """
    logger.info(f"Building project for problem: {data.problem_statement}")
    
    return {
        "status": "success",
        "project": {
            "problem": data.problem_statement,
            "solution": "Low-cost water filter using charcoal, sand, and pebbles.",
            "plan_steps": [
                "Gather materials (plastic bottle, charcoal, sand).",
                "Cut the bottle in half.",
                "Layer the materials.",
                "Test with muddy water."
            ],
            "impact_metrics": ["Litres of water purified per day", "Cost per filter"]
        }
    }


@router.post("/voice-report")
async def generate_voice_report(data: VoiceReportInput):
    """
    Parent Voice Reports
    Generates a text summary in Marathi, which would normally be passed to a TTS engine (like Google/AWS).
    """
    logger.info("Generating parent voice report text")
    
    text = (
        f"नमस्कार. या आठवड्यात तुमच्या {data.student_name} ने "
        f"{data.lessons_completed} धडे पूर्ण केले. "
        f"{data.skills_learned} नवीन कौशल्ये शिकली. "
        f"उपस्थिती {data.attendance_percent} टक्के आहे. "
        f"अभिनंदन."
    )
    
    return {
        "status": "success",
        "report_text_marathi": text,
        "audio_url": None # In a real implementation, this would be an S3 URL to the generated MP3
    }

@router.post("/teacher-assistant", response_model=TeacherAssistantResponse)
async def teacher_assistant_endpoint(request: TeacherAssistantRequest):
    """
    Teacher Assistant
    Generates lesson plans, quizzes, and activities.
    """
    logger.info(f"Generating teacher assistant response for {request.task}")
    try:
        return await generate_teacher_assistant_response(request)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: services/api/src/routes/intelligence.ts
import { FastifyInstance } from 'fastify';

export default async function intelligenceRoutes(fastify: FastifyInstance) {
  const authenticate = { preHandler: [fastify.authenticate] };

  // =============================================================================
  // PHASE 3: LEARNING INTELLIGENCE ENGINE & KNOWLEDGE GRAPH
  // =============================================================================

  // GET /v1/intelligence/learning-index
  fastify.get('/learning-index', authenticate, async (request, reply) => {
    const { studentId } = request.query as { studentId: string };
    if (!studentId) return reply.code(400).send({ success: false, message: "studentId required" });

    // 1. Fetch Mastery
    const masteries = await fastify.prisma.studentConceptMastery.findMany({ where: { studentId } });
    const avgMastery = masteries.length ? masteries.reduce((acc, m) => acc + m.masteryScore, 0) / masteries.length : 0;
    
    // 2. Fetch Quiz Performance
    const progress = await fastify.prisma.studentProgress.findMany({ where: { studentId } });
    const avgQuizScore = progress.length ? progress.reduce((acc, p) => acc + (p.avgScore || 0), 0) / progress.length : 0;
    
    // 3. Activity & Project Mocks (Since we don't have explicit activity completion tracking yet)
    const activityCompletionRate = 75; // Mock
    const projectCompletionRate = 50; // Mock

    // Formula: (Avg Concept Mastery %) * 0.4 + (Avg Quiz Score) * 0.3 + (Activity Rate) * 0.15 + (Project Rate) * 0.15
    const learningIndex = (avgMastery * 0.4) + (avgQuizScore * 0.3) + (activityCompletionRate * 0.15) + (projectCompletionRate * 0.15);

    return reply.send({
      success: true,
      data: {
        learningIndex: Math.round(learningIndex),
        skillReadinessIndex: 45,
        careerReadinessIndex: 60,
        innovationIndex: 30,
        communityImpactIndex: 50,
      }
    });
  });

  // GET /v1/intelligence/recommendations
  fastify.get('/recommendations', authenticate, async (request, reply) => {
    const { studentId } = request.query as { studentId: string };
    
    // Graph Traversal Logic
    // Find weak concepts -> Map to activities/quizzes.
    const weakMasteries = await fastify.prisma.studentConceptMastery.findMany({
      where: { studentId, status: { in: ['NOT_STARTED', 'LEARNING'] } },
      include: { concept: true },
      take: 3
    });

    const recommendations = weakMasteries.map(wm => ({
      type: "CONCEPT_PRACTICE",
      title: `Practice ${wm.concept.name}`,
      reason: "Weak concept detected.",
      actionUrl: `/learning-os/concept/${wm.conceptId}`
    }));

    // Add career/village recommendations based on interests (Graph Traversal mock)
    recommendations.push({
      type: "VILLAGE_PROJECT",
      title: "Water Conservation Model",
      reason: "Matches your interest in Agriculture.",
      actionUrl: `/village/projects`
    });

    return reply.send({ success: true, data: recommendations });
  });

  // GET /v1/intelligence/career-dna
  fastify.get('/career-dna', authenticate, async (request, reply) => {
    const { studentId } = request.query as { studentId: string };

    let dna = await fastify.prisma.careerDNAProfile.findUnique({ where: { studentId } });
    if (!dna) {
      // Create mock DNA profile if not exists
      dna = await fastify.prisma.careerDNAProfile.create({
        data: {
          studentId,
          interests: ["Technology", "Environment"],
          topSkillIds: [],
          calculatedCareers: ["Software Engineer", "Environmental Scientist", "Agritech Founder", "Data Analyst", "Civil Engineer"]
        }
      });
    }

    return reply.send({
      success: true,
      data: {
        interests: dna.interests,
        topCareers: dna.calculatedCareers
      }
    });
  });

  // POST /v1/intelligence/mastery/update (Matches user requirement)
  fastify.post('/mastery/update', authenticate, async (request, reply) => {
    const { studentId, conceptId, status, masteryScore } = request.body as any;
    
    const mastery = await fastify.prisma.studentConceptMastery.upsert({
      where: { studentId_conceptId: { studentId, conceptId } },
      update: { status, masteryScore },
      create: { studentId, conceptId, status, masteryScore }
    });

    // Check Chapter Hard Gate Logic
    const chapter = await fastify.prisma.concept.findUnique({
      where: { id: conceptId },
      select: { chapterId: true }
    });

    const allConceptsInChapter = await fastify.prisma.concept.findMany({
      where: { chapterId: chapter!.chapterId }
    });

    const studentMasteries = await fastify.prisma.studentConceptMastery.findMany({
      where: { studentId, conceptId: { in: allConceptsInChapter.map(c => c.id) } }
    });

    const isChapterMastered = allConceptsInChapter.every(c => {
      const m = studentMasteries.find(sm => sm.conceptId === c.id);
      return m && m.status === 'MASTERED';
    });

    return reply.send({
      success: true,
      data: {
        mastery,
        isChapterMastered,
        message: isChapterMastered ? "Chapter unlocked!" : "Keep practicing to unlock the next chapter."
      }
    });
  });

  fastify.post('/learning-recovery', async (request, reply) => {
    // 1. Fetch student performance from DB
    // 2. Call AI Service: POST http://vidyasetu-ai:8000/api/v1/intelligence/learning-recovery
    // 3. Save recovery plan to LearningIntelligenceProfile
    
    return {
      success: true,
      message: "Adaptive learning recovery plan generated.",
      data: {
        weak_concepts: ["Fractions", "Decimals"],
        strong_concepts: ["Addition", "Multiplication"],
        learning_velocity: 0.8,
        recovery_plan: {
          daily_missions: ["Practice 10 Fractions questions", "Practice 10 Decimals questions"],
          recommended_lessons: ["Basic Fractions Video", "Decimal Division Interactive"]
        }
      }
    };
  });

  fastify.post('/daily-briefing', async (request, reply) => {
    // 1. Fetch student's 5-pillar Growth Engine stats from frontend store (via body)
    // 2. Call AI Service: POST http://vidyasetu-ai:8000/api/v1/intelligence/daily-briefing
    
    return {
      success: true,
      message: "Daily briefing generated.",
      data: {
        briefing_text: "नमस्कार! तू सध्या Growing लेव्हल वर आहेस. आज आपण 'कौशल्ये (Skills)' मध्ये प्रगती करूया. आजची मिशन पूर्ण करून तुझी लेव्हल वाढव!",
        focus_area: "कौशल्ये (Skills)"
      }
    };
  });

  fastify.post('/career-match', async (request, reply) => {
    // 1. Fetch Student's Learning & Skill scores
    // 2. Call AI Service
    // 3. Save to DreamCareer table
    
    return {
      success: true,
      data: {
        career_matches: [
          {
            career: "Agriculture Scientist",
            match_score: 92,
            reason: "Strong interest in nature and high science learning scores."
          }
        ]
      }
    };
  });

  fastify.post('/project-builder', async (request, reply) => {
    // 1. Take problem from student
    // 2. Call AI Service
    // 3. Save to AIProject table
    
    return {
      success: true,
      data: {
        project: {
          problem: "Water Problem",
          solution: "Low-cost water filter using charcoal, sand, and pebbles.",
          plan_steps: ["Gather materials", "Cut bottle", "Test"],
          impact_metrics: ["Litres purified"]
        }
      }
    };
  });

  fastify.post('/voice-report', async (request, reply) => {
    try {
      const aiUrl = process.env.AI_SERVICE_URL || 'http://localhost:8000';
      const jwtToken = fastify.jwt.sign({ service: 'vidyasetu-api' });

      const response = await fetch(`${aiUrl}/api/v1/intelligence/voice-report`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${jwtToken}`
        },
        body: JSON.stringify(request.body)
      });

      if (!response.ok) throw new Error(`AI Service returned ${response.status}`);
      const data = await response.json();
      
      return { success: true, data };
    } catch (error) {
      request.log.error(error);
      return reply.status(500).send({ success: false, message: "Voice Report service is unavailable." });
    }
  });

  // AI Tutor Proxy
  fastify.post('/tutor/chat', { preHandler: [fastify.authenticate] }, async (request, reply) => {
    try {
      const aiUrl = process.env.AI_SERVICE_URL || 'http://localhost:8000';
      const jwtToken = fastify.jwt.sign({ service: 'vidyasetu-api' });

      const response = await fetch(`${aiUrl}/api/v1/tutor/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${jwtToken}`
        },
        body: JSON.stringify(request.body)
      });

      if (!response.ok) {
        throw new Error(`AI Service returned ${response.status}`);
      }

      const data = await response.json();
      return reply.send(data);
    } catch (error) {
      request.log.error(error);
      return reply.status(500).send({ success: false, message: "AI Mentor service is unavailable." });
    }
  });

  // Teacher Assistant Proxy
  fastify.post('/teacher-assistant', { preHandler: [fastify.authenticate] }, async (request, reply) => {
    try {
      const aiUrl = process.env.AI_SERVICE_URL || 'http://localhost:8000';
      const jwtToken = fastify.jwt.sign({ service: 'vidyasetu-api' });

      const response = await fetch(`${aiUrl}/api/v1/intelligence/teacher-assistant`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${jwtToken}`
        },
        body: JSON.stringify(request.body)
      });

      if (!response.ok) {
        throw new Error(`AI Service returned ${response.status}`);
      }

      const data = await response.json();
      return reply.send(data);
    } catch (error) {
      request.log.error(error);
      return reply.status(500).send({ success: false, message: "Teacher Assistant service is unavailable." });
    }
  });

  // Quiz Generator Proxy
  fastify.post('/generate-quiz', { preHandler: [fastify.authenticate] }, async (request, reply) => {
    try {
      const aiUrl = process.env.AI_SERVICE_URL || 'http://localhost:8000';
      const jwtToken = fastify.jwt.sign({ service: 'vidyasetu-api' });

      const response = await fetch(`${aiUrl}/api/v1/tutor/generate-quiz`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${jwtToken}`
        },
        body: JSON.stringify(request.body)
      });

      if (!response.ok) {
        throw new Error(`AI Service returned ${response.status}`);
      }

      const data = await response.json();
      return reply.send(data);
    } catch (error) {
      request.log.error(error);
      return reply.status(500).send({ success: false, message: "Quiz Generator service is unavailable." });
    }
  });
}
